src/Cache/Cache.php

```diff
@@ -0,0 +1,121 @@
+        CacheInterface::LONG => 7200,
+        CacheInterface::X_LONG => 86400,
+        CacheInterface::INDEFINITE => 0,
+    ];
+
+    /** @var array */
+    private $cacheTimes;
+
+    public function __construct(
+        CacheItemPoolInterface $cachePool,
+        string $prefix,
+        array $cacheTimes = []
+    ) {
+        $this->cachePool = $cachePool;
+        $this->prefix = $prefix;
+        $this->cacheTimes = array_merge($this->defaultCacheTimes, $cacheTimes);
+    }
+
+    public function getItem(string $key): CacheItemInterface
+    {
+        $key = $this->standardiseKey($key);
+        if ($this->flushCacheItems) {
+            $this->cachePool->deleteItem($key);
+        }
+        return $this->cachePool->getItem($key);
+    }
+
+    public function setItem(CacheItemInterface $item, $ttl): bool
+    {
+        $ttl = $this->calculateTtl($ttl);
+        $item->expiresAfter($ttl);
+        return $this->cachePool->save($item);
+    }
+
+    /**
+     * IF CALLABLE RETURNS SOMETHING THAT EVALUATES TO EMPTY THE RESULT WILL NOT BE CACHED
+     *
+     * @param string $key
+     * @param int|class constant $ttl
+     * @param callable $function
+     * @param array|null $arguments
+     * @return mixed
+     */
+    public function getOrSet(string $key, $ttl, callable $function, array $arguments = [])
+    {
+        $cacheItem = $this->getItem($key);
+        if ($cacheItem->isHit() && !$this->flushCacheItems) {
+            return $cacheItem->get();
+        }
+        $ttl = $this->calculateTtl($ttl);
+        $result = call_user_func_array($function, $arguments);
+        if (!empty($result)) {
+            $cacheItem->set($result);
+            $cacheItem->expiresAfter($ttl);
+            $this->cachePool->save($cacheItem);
+        }
+        return $result;
+    }
+
+    public function keyHelper(string $className, string $functionName, string ...$uniqueValues): string
+    {
+        // Please help prevent cache namespace collisions by driving carefully
+        $uniqueValues = str_replace('.', '_', $uniqueValues);
+        $values = [$className, $functionName] + $uniqueValues;
+        return join('.', $values);
+    }
+
+    public function setFlushCacheItems(bool $flushCacheItems): void
+    {
+        $this->flushCacheItems = $flushCacheItems;
+    }
+
+    private function standardiseKey(string $key)
+    {
+        $key = $this->prefix . '.' . $key;
+        return preg_replace('/[^A-Za-z0-9_\.]/', '_', $key);
+    }
+
+    private function calculateTtl($ttl): int
+    {
+        if (is_numeric($ttl)) {
+            $ttl = (int) $ttl;
+        } elseif (is_string($ttl) && isset($this->cacheTimes[$ttl])) {
+            $ttl = $this->cacheTimes[$ttl];
+        } else {
+            throw new \InvalidArgumentException("TTL must be a number or Cache class constant");
+        }
+        return $this->protectLifetimeFromStampede($ttl);
+    }
+
+    private function protectLifetimeFromStampede(int $ttl): int
```

Comment:
I know what this does. But someone in 6 months may not, should this have a docblock?